votor/src/consensus_pool/vote_pool.rs

```diff
@@ -1,129 +1,304 @@
+//! Container to store received votes and associated stakes.
+//!
+//! Also implements various checks for slashable behavior and duplicate votes.
+
 use {
-    crate::common::Stake,
-    agave_votor_messages::consensus_message::VoteMessage,
+    crate::common::{Stake, VoteType},
+    agave_votor_messages::{consensus_message::VoteMessage, vote::Vote},
+    solana_clock::Slot,
     solana_hash::Hash,
     solana_pubkey::Pubkey,
-    std::collections::{BTreeMap, BTreeSet},
+    std::collections::{btree_map::Entry, BTreeMap},
+    thiserror::Error,
 };
 
-/// There are two types of vote pools:
-/// - SimpleVotePool: Tracks all votes of a specfic vote type made by validators for some slot N, but only one vote per block.
-/// - DuplicateBlockVotePool: Tracks all votes of a specfic vote type made by validators for some slot N,
-///   but allows votes for different blocks by the same validator. Only relevant for VotePool's that are of type
-///   Notarization or NotarizationFallback
-pub(super) enum VotePool {
-    SimpleVotePool(SimpleVotePool),
-    DuplicateBlockVotePool(DuplicateBlockVotePool),
+#[derive(Debug, PartialEq, Eq, Error)]
+pub(crate) enum AddVoteError {
+    #[error("duplicate vote")]
+    Duplicate,
+    #[error("slashable behavior")]
+    Slash,
 }
 
-#[derive(Default)]
-pub(super) struct SimpleVotePool {
-    votes: Vec<VoteMessage>,
-    total_stake: Stake,
-    prev_voted_validators: BTreeSet<Pubkey>,
+/// Container to store per slot votes.
+struct Votes {
```

Comment:
I don't think Votes is a good name here, Votes seems to be a vector of VoteMessage while this is structured storage for votes. What's wrong with VotePool?

Reply:
renamed.